### ISSUE TITLE
cx_Oracle >=6.0.3 doesn't allow to use NaNs 

### ISSUE THREAD

**Issue by leonid-butenko** [oracle/python-cx_Oracle#184]
I found here oracle/python-cx_Oracle#91 that the usage of NaN values corrupts the data. We've been using NaNs for a few years now and with a recent update of cx_Oracle have realized it's not supported any longer. What are the alternatives for NaNs? From scientific point of view, having NaNs is quite essential and important.

here is the code snippet I use to reproduce the issue:

```
import cx_Oracle

c = cx_Oracle.connect(
    dsn=cx_Oracle.makedsn("db", 1521, sid="TEST"), user="USER", password="PWD")

cursor = c.cursor()

try:
    cursor.execute("drop table foo")
except:
    pass

cursor.execute("""
    create table foo(
        id integer primary key,
        data BINARY_FLOAT
    )
""")

cursor.execute("INSERT INTO foo (id, data) VALUES (:id, :data)", {"id": 1, "data": float(100.0)})
cursor.execute("UPDATE foo SET data=:data", {"data": float('nan')})

c.commit()
```

I get the following error:

```
>python ./test.py 
Traceback (most recent call last):
  File "./test.py", line 21, in <module>
    cursor.execute("UPDATE foo SET data=:data", {"data": float('nan')})
cx_Oracle.DatabaseError: DPI-1055: value is not a number (NaN) and cannot be used in Oracle numbers

```
#### Answer the following questions:

1. What is your version of Python? Is it 32-bit or 64-bit?

python 2.7.14, 64 bits

2. What is your version of cx_Oracle?

6.3

3. What is your version of the Oracle client (e.g. Instant Client)?  How was it
   installed?  Where is it installed?

I tried with 11.2 and 12.2

4. What is your version of the Oracle Database?

11g enterprise

5. What is your OS and version?

ubuntu 18.04 and Debian 9

6. What compiler version did you use?  For example, with GCC, run
   `gcc --version`.

gcc (GCC) 4.8.5
Copyright (C) 2015 Free Software Foundation, Inc.

7. What environment variables did you set?  How *exactly* did you set them?

I use anaconda cloud running in docker. 

8. What *exact* command caused the problem (e.g. what command did you try to
   install with)?  Who were you logged in as?
```
>docker run -it --rm continuumio/miniconda bash
>conda install -c conda-forge -c anaconda  cx_oracle=6.0.3 oracle-instantclient libaio
>python test.py
```
9. What error(s) you are seeing?

> Traceback (most recent call last):
>   File "./test.py", line 21, in <module>
>     cursor.execute("UPDATE foo SET data=:data", {"data": float('nan')})
> cx_Oracle.DatabaseError: DPI-1055: value is not a number (NaN) and cannot be used in Oracle numbers

**Comment by zzzeek** [comment]
Oracle's documentation implies NaN is supported:  https://docs.oracle.com/cd/E11882_01/server.112/e41084/sql_elements001.htm#SQLRF50979

this answer https://stackoverflow.com/a/28461506/34549 seems to indicate you need to use binary_float or binary_double datatypes.   should that work ?

**Comment by leonid-butenko** [comment]
Thanks for finding links! Pls note, I use BINARY_FLOAT in the snippet above.

**Comment by zzzeek** [comment]
Well, good thing I added logic to allow more specific setinputsizes setup for specific SQLAlchemy sub-types.

What are the downsides to NATIVE_FLOAT? I need to know when I'm using this and when I'm not.  Again, we go to https://github.com/oracle/python-cx_Oracle/issues/68#issuecomment-323451985, where I changed SQLAlchemy to not setinputsizes() for numeric types.     Especially, if the datatype is BINARY_DOUBLE, and in an executemany() there are lots of wide-precision numeric values coming in, how do I set NATIVE_FLOAT there without introducing truncation for the non-NaN values ?

**Comment by zzzeek** [comment]
ah you know I was confusing DOUBLE with NUMBER.  per https://docs.oracle.com/cd/B28359_01/server.111/b28285/sqlqr06.htm#i76716, BINARY_FLOAT and BINARY_DOUBLE are both floating point types.  but...what is the precision we expect with these two types?    my goal is to avoid the bug in https://github.com/oracle/python-cx_Oracle/issues/68.   I need numeric values to round-trip perfectly given the precision that the type defines.

so far it seems like I should setinputsizes() to NATIVE_FLOAT with the BINARY_FLOAT and BINARY_DOUBLE types.  I don't care about a slight performance penalty.  I care about accuracy being maintained and no surprise truncation.

**Comment by anthony-tuininga** [comment]
BINARY_FLOAT is single precision floating point and BINARY_DOUBLE is double precision floating point (the same type that Python uses internally for its "float" type). So you will get the same precision as those values allow. See [here](https://en.wikipedia.org/wiki/Double-precision_floating-point_format) for all the nitty gritty details, but the main point for double is you get 2^53 as the maximum integer that can be used and about 15 significant digits and a maximum value of 10^308. Oracle's NUMBER format has a far smaller scale but greater precision. You can typically get 38 significant digits and all of those can be used to represent an integer but the maximum value is only 10^126.

Given those constraints the input from Python and the target in the database are both important. If you have BINARY_FLOAT in the database you have to ensure that the input data is Python floats, integers or decimal values that will fit inside the constraints of BINARY_FLOAT. If you don't do so, you will get silent truncation. The same will be true for BINARY_DOUBLE. Clearly Python floats will always fit perfectly inside BINARY_DOUBLE but integers and decimals can easily exceed its constraints. For NUMBER data Python floats will always result in some truncation for some numbers. Most of the time that isn't an issue but its always there because NUMBER is a decimal format, not a binary format (unlike BINARY_FLOAT and BINARY_DOUBLE which are binary format). For that reason Python decimals and integers fit perfectly into Oracle NUMBER but again both integers and decimals can easily exceed the constraints of NUMBER. And for the same reason Python decimals will always result in conversion errors when going into binary format.

If it sounds complicated, that's because it is! Thankfully most of the time all of these "issues" can be ignored as the precision/scale that most people are dealing with are well within the constraints of all of these types. But if you deal with numbers of larger precision/scale and values like NaN on a regular basis then you have to know all of this stuff and know how to make the software behave the way it should. cx_Oracle has the ability to do this using cursor.setinputsizes() and tries its best to use reasonable defaults. There is no way that I am aware of to do it perfectly without programmer input -- but I'm interested in anything you might suggest to improve things!

**Comment by zzzeek** [comment]
Hi Anthony -

OK, I have Python float("nan") round tripping.  But now what about the same thing #78 for INFINITY?   Decimal("NaN") should round trip as well.  I'm calling setinputsizes(NATIVE_FLOAT):

`cx_Oracle.DatabaseError) DPI-1043: invalid number [SQL: 'INSERT INTO t1 (intcol, numericcol) VALUES (:intcol, :numericcol)'] [parameters: [{'intcol': 1, 'numericcol': Decimal('NaN')}, {'intcol': 2, 'numericcol': Decimal('NaN')}]]`

**Comment by anthony-tuininga** [comment]
@zzzeek, I have just pushed a commit that will permit binding decimal values to cx_Oracle.NATIVE_FLOAT. Truncation will occur, of course, if the decimal value exceeds the constraints of a double value, but since you are binding to cx_Oracle.NATIVE_FLOAT I think that makes sense. This is not permitted for cx_Oracle.NUMBER however, to avoid the problem noted in issue #68.

@leonid-butenko, apparently that was a regression in 6.2 which was corrected in 6.3 (and a test case added to the test suite to prevent it from happening again).
